mujoco_py/tests/test_substep.py

```diff
@@ -0,0 +1,164 @@
+        <position joint="j2" kp="100"/>
+    </actuator>
+</mujoco>
+'''
+
+HELLO_FN = '''
+    #include <stdio.h>
+    void fun(const mjModel* m, mjData* d) {
+        printf("hello");
+    }
+'''
+
+INCREMENT_FN = '''
+    void fun(const mjModel* m, mjData* d) {
+        d->userdata[0] += 1;
+    }
+'''
+
+
+class TestSubstep(unittest.TestCase):
+    def test_hello(self):
+        sim = MjSim(load_model_from_xml(XML.format(nuserdata=0)))
+        sim.set_substep_callback(HELLO_FN)
+        sim.step()  # should print 'hello'
+
+    def test_increment(self):
+        sim = MjSim(load_model_from_xml(XML.format(nuserdata=1)))
+        sim.set_substep_callback(INCREMENT_FN)
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.step()  # should increment userdata[0]
+        self.assertEqual(sim.data.userdata[0], 1)
+        # Test again with different nsubsteps, reuse model
+        sim = MjSim(sim.model, nsubsteps=7)
+        sim.set_substep_callback(INCREMENT_FN)
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.step()  # should increment userdata[0] 7 times
+        self.assertEqual(sim.data.userdata[0], 7)
+
+    def test_sum_ctrl(self):
+        fn = '''
+            void fun(const mjModel* m, mjData* d) {
+                for (int i = 0; i < m->nu; i++) {
+                    my_sum += d->ctrl[i];
+                }
+            }
+        '''
+        model = load_model_from_xml(XML.format(nuserdata=1))
+        model.set_userdata_names(['my_sum'])
+        sim = MjSim(model)
+        sim.set_substep_callback(fn)
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.step()  # should set userdata[0] to sum of controls
+        self.assertEqual(sim.data.userdata[0], 0)
+        sim.data.ctrl[0] = 12.
+        sim.data.ctrl[1] = .34
+        sim.step()
+        self.assertEqual(sim.data.userdata[0], 12.34)
+        sim.step()
+        self.assertEqual(sim.data.userdata[0], 24.68)
+
+    def test_penetrations(self):
+        # Test that we capture the max penetration of a falling sphere
+        # Given a single step with many substeps
+        xml = '''
+            <mujoco>
+                <size nuserdata="1"/>
+                <worldbody>
+                    <body pos="0 0 1">
+                        <joint type="free"/>
+                        <geom size=".1"/>
+                    </body>
+                    <geom type="plane" size="1 1 1"/>
+                </worldbody>
+            </mujoco>
+        '''
+        # NOTE: penetration is negative distance in contacts
+        fn = '''
+            #define MIN(a, b) (a < b ? a : b)
+            void fun(const mjModel* m, mjData* d) {
+                for (int i = 0; i < d->ncon; i++) {
+                    min_contact_dist = MIN(min_contact_dist, d->contact[i].dist);
+                }
+            }
+        '''
+        fields = ['min_contact_dist']
+        model = load_model_from_xml(xml)
+        model.set_userdata_names(fields)
```

Comment:
could we have one function on sim ? instead of set_userdata_names on model, and set_substep_callback on sim ??

Something like:
sim. set_substep_callback(fun, fields)

Reply:
done